File: c++/51 - N-Queens.cpp
// 51
// N-Queens
// https://leetcode.com/problems/n-queens/
// c++
// hard
// O(n^n)
// O(n)
// basic

// time: O(n^n)
// space: O(n)
class Solution {
public:
    bool valid(int& r, int& c, vector<int>& vec) {  
        for (int i = 0; i < vec.size(); i++) {
            if (vec[i] == c || abs(vec[i]-c) == abs(i-r))
                return false;
        }
        return true;
    }
    
    void visit(int& n, int r, vector<int> vec, vector<vector<string>>& results) {        
        for (int c = 0; c < n; c++) {
            if (!valid(r, c, vec))
                continue;
            if (r == n-1) {
                vector<string> result;
                for (auto p : vec) {
                    string s(n, '.');
                    s.replace(p, 1, "Q");
                    result.push_back(s);
                }
                string s(n, '.');
                s.replace(c, 1, "Q");
                result.push_back(s);
                results.push_back(result);
            }    
            else {
                vec.push_back(c);
                visit(n, r+1, vec, results);
                vec.pop_back();
            }
        }
    }
    
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> results;
        vector<int> vec;
        visit(n, 0, vec, results);
        return results;
    }
};

File: python/51 - N-Queens.py
# 51
# N-Queens
# https://leetcode.com/problems/n-queens/
# python
# hard
# O(n!)
# O(n)
# matrix


class Solution:
    def valid(self, r, c, queens):
        for x, y in queens:
            if c == y or abs(r - x) == abs(c - y):
                return False
        return True

    def visit(self, r, queens):
        for c in range(self.n):
            if not self.valid(r, c, queens):
                continue
            elif r == self.n - 1:
                self.ans.append(
                    [
                        "." * y + "Q" + "." * (self.n - y - 1)
                        for _, y in queens[:] + [(r, c)]
                    ]
                )
            else:
                self.visit(r + 1, queens[:] + [(r, c)])

    def solveNQueens(self, n: int) -> List[List[str]]:
        self.n = n
        self.ans = []
        self.visit(0, [])
        return self.ans if n > 0 else [[]]
